lib/internal/test_runner/coverage.js

```diff
@@ -289,6 +293,116 @@ class TestCoverage {
       }
     }
   }
+
+  getCoverageFromDirectory() {
+    const result = new SafeMap();
+    let dir;
+
+    try {
+      dir = opendirSync(this.coverageDirectory);
+
+      for (let entry; (entry = dir.readSync()) !== null;) {
+        if (RegExpPrototypeExec(kCoverageFileRegex, entry.name) === null) {
+          continue;
+        }
+
+        const coverageFile = join(this.coverageDirectory, entry.name);
+        const coverage = JSONParse(readFileSync(coverageFile, 'utf8'));
+        mergeCoverage(result, this.mapCoverageWithSourceMap(coverage));
+      }
+
+      return ArrayFrom(result.values());
+    } finally {
+      if (dir) {
+        dir.closeSync();
+      }
+    }
+  }
+
+
+  mapCoverageWithSourceMap(coverage) {
+    const { result } = coverage;
+    const sourceMapCache = coverage['source-map-cache'];
+    if (!sourceMapCache) {
+      return result;
+    }
+    const newResult = new SafeMap();
+    for (let i = 0; i < result.length; ++i) {
+      const script = result[i];
+      const { url, functions } = script;
+
+      if (shouldSkipFileCoverage(url) || sourceMapCache[url] == null) {
+        newResult.set(url, script);
+        continue;
+      }
+      const originalLines = this.getLines(url);
+      const { data, lineLengths } = sourceMapCache[url];
+      if (data.sourcesContent != null) {
+        for (let j = 0; j < data.sources.length; ++j) {
+          this.getLines(data.sources[j], data.sourcesContent[j]);
+        }
+      }
+      const sourceMap = new SourceMap(data, { __proto__: null, lineLengths });
+
+      for (let j = 0; j < functions.length; ++j) {
+        const { ranges, functionName, isBlockCoverage } = functions[j];
+        if (ranges == null) {
+          continue;
+        }
+        let newUrl;
+        const newRanges = [];
+        for (let k = 0; k < ranges.length; ++k) {
+          const { startOffset, endOffset, count } = ranges[k];
+          const { lines } = mapRangeToLines(ranges[k], originalLines);
+
+          let startEntry = sourceMap
+            .findEntry(lines[0].line - 1, MathMax(0, startOffset - lines[0].startOffset));
+          const endEntry = sourceMap
+            .findEntry(lines[lines.length - 1].line - 1, (endOffset - lines[lines.length - 1].startOffset) - 1);
+          if (!startEntry.originalSource && endEntry.originalSource &&
+            lines[0].line === 1 && startOffset === 0 && lines[0].startOffset === 0) {
+            // Edge case when the first line is not mappable
```

Comment:
Why is the first line not mappable? It looks like there are a couple more "not mappable" comments. Do you know why they aren't mappable? If so, can it be included in the comments.

Reply:
what `findEntry` does is a binary search for the first entry that appears in the source map **after** the line/column you request. 
there are many cases where such an entry cannot be found (generated code doesn't always point to a source for various reasons - it might be a comment added by the build tool source maps exclude `node_modules`, etc)
this specific edge case is important since v8 coverage will almost always have a range starting at the beginning of generating code (`{line:0,col:0}`) until the end of the end of the file. since this range marks the count of for all top-level lines outside of a functions it is very important so I added a specific handling for it

https://github.com/nodejs/node/blob/1abff073921bcb0631602032aef0135bccfaee0d/lib/internal/source_map/source_map.js#L190-L220